layer: fix same rw layer races

kolyshkin opened the pull request:
_(this is a fix to a bug introduced in #39135, and also a followup to #38265)_

Consider the following scenario:

```
----- goroutine 1 -----               ----- goroutine 2 -----
ReleaseRWLayer()
  m := ls.mounts[l.Name()]
  ...
  m.deleteReference(l)
  m.hasReferences()
  ...                                 GetRWLayer()
  ...                                   mount := ls.mounts[id]
  ls.driver.Remove(m.mountID)
  ls.store.RemoveMount(m.name)          return mount.getReference()
  delete(ls.mounts, m.Name())
-----------------------               -----------------------
```

When something like this happens, `GetRWLaye`r will return
an `RWLayer` without a storage. Oops.

There might be more races like this, and it seems the best
solution is to lock by layer id/name by using pkg/locker.
    
With this in place, name collision could not happen, so remove
the part of previous commit that protected against it in
CreateRWLayer (temporary nil assigmment and associated rollback).
    
So, now we have
    * layerStore.mountL sync.Mutex to protect layerStore.mount map[]
      (against concurrent access);
    * mountedLayer's embedded `sync.Mutex` to protect its references map[];
    * layerStore.layerL (which I haven't touched);
    * per-id locker, to avoid name conflicts and concurrent operations
      on the same rw layer.
    
The whole rig seems to look more readable now (mutexes use is
straightforward, no nested locks).

kolyshkin commented:
@dmcgowan @tonistiigi PTAL

kolyshkin commented:
CI failures are def undelated

kolyshkin committed:
layer/CreateRWLayerByGraphID: remove

This is an additon to commit 1fea38856a ("Remove v1.10 migrator")
aka PR #38265. Since that one, CreateRWLayerByGraphID() is not
used anywhere, so let's drop it.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@AkihiroSuda PTAL (first commit)

dmcgowan reviewed (approved):
LGTM

tonistiigi commented on layer/layer_store.go:
```diff
@@ -583,6 +585,10 @@ func (ls *layerStore) GetMountID(id string) (string, error) {
 	ls.mountL.Lock()
 	m := ls.mounts[l.Name()]
+	if m != nil {
+		m.Lock()
```
Should `RWLayer` be called one too many times(eg. refcount gets confused from error handling or live-restore) compared to the ref count then this seems like a deadlock. The first request will be stuck on `driver.Remove` second one will come in and block on `m.Lock`. After `driver.Remove` finishes it can't continue cause the second one is holding the `ls.mountL`.

tonistiigi replied:
nit: on contention for the same layer (eg. last refs released together that is quite normal) this will still block the whole mountL. Better than before but not as good as what pkg/locker and similar packages do.

kolyshkin replied:
I wish my built-in race detector would be as good as yours ;) Thanks!

I've replaced the ugly hack of mis-using `mountedLayer.Lock()` to per-id lock. Now we have
* `mountL sync.Mutex` which is used solely for concurrent access to `layerStore.mount map[]`
* `mountedLayer sync.Mutex` which is used solely for concurrent acccess to `mountedLayer.references map[]`
* per-id locker, to avoid name conflicts and concurrent ops on the same layer.

The whole rig seems to look more readable now (mutexes use is straightforward, no nested locks).

@tonistiigi PTALA

kolyshkin committed:
layer: protect from same-name races

As pointed out by Tonis, there's a race between ReleaseRWLayer()
and GetRWLayer():

```
----- goroutine 1 -----               ----- goroutine 2 -----
ReleaseRWLayer()
  m := ls.mounts[l.Name()]
  ...
  m.deleteReference(l)
  m.hasReferences()
  ...                                 GetRWLayer()
  ...                                   mount := ls.mounts[id]
  ls.driver.Remove(m.mountID)
  ls.store.RemoveMount(m.name)          return mount.getReference()
  delete(ls.mounts, m.Name())
-----------------------               -----------------------
```

When something like this happens, GetRWLayer will return
an RWLayer without a storage. Oops.

There might be more races like this, and it seems the best
solution is to lock by layer id/name by using pkg/locker.

With this in place, name collision could not happen, so remove
the part of previous commit that protected against it in
CreateRWLayer (temporary nil assigmment and associated rollback).

So, now we have
* layerStore.mountL sync.Mutex to protect layerStore.mount map[]
  (against concurrent access);
* mountedLayer's embedded `sync.Mutex` to protect its references map[];
* layerStore.layerL (which I haven't touched);
* per-id locker, to avoid name conflicts and concurrent operations
  on the same rw layer.

The whole rig seems to look more readable now (mutexes use is
straightforward, no nested locks).

Reported-by: Tonis Tiigi <tonistiigi@gmail.com>
Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
@tonistiigi @dmcgowan PTALA (this one should be way easier to review)

kolyshkin commented:
1. experimental CI fails on TestPortExposeHostBinding, I'm seeing this for the first time.

```
00:42:23.397 ----------------------------------------------------------------------
00:42:23.397 FAIL: docker_cli_port_test.go:308: DockerSuite.TestPortExposeHostBinding
00:42:23.397 
00:42:23.397 docker_cli_port_test.go:327:
00:42:23.397     // Port is still bound after the Container is removed
00:42:23.397     c.Assert(err, checker.NotNil, check.Commentf("out: %s", out))
00:42:23.397 ... value = nil
00:42:23.397 ... out: 
```

Not sure why it is failing.

2. janky and powerpc fails on
```
00:52:10.813 FAIL: docker_cli_search_test.go:47: DockerSuite.TestSearchCmdOptions
00:52:10.813 
00:52:10.815 assertion failed: expression is false: strings.Count(outSearchCmdStars1, "[OK]") <= strings.Count(outSearchCmd, "[OK]"): The quantity of images with stars should be less than that of all images: NAME                        DESCRIPTION                                     STARS     OFFICIAL   AUTOMATED
00:52:10.815 busybox                     Busybox base image.                             1584      [OK]       
00:52:10.815 progrium/busybox                                                            69                   [OK]
....
```

According to https://github.com/moby/moby/issues/26633, it could mean problems with Docker Hub. Still, test can be improved, see https://github.com/moby/moby/pull/39243

thaJeztah commented on layer/layer_store.go:
```diff
@@ -36,6 +37,8 @@ type layerStore struct {
 	mounts map[string]*mountedLayer
 	mountL sync.Mutex
+
+	locker *locker.Locker
```
it's common to put locks above the fields they are protecting (perhaps a comment would help as well, describing what the locks are for)

kolyshkin replied:
I did had a comment in some version :) but it got lost in the turmoil. Fixed.

kolyshkin commented:
Indeed we have problems with Docker Hub which affect our CI. This is from the latest run on janky:

```
02:04:45.307 E   APIError: 500 Server Error: Internal Server Error ("Get https://registry-1.docker.io/v2/: EOF")
02:04:45.307 _________________ ERROR at setup of ExecTest.test_exec_inspect _________________
02:04:45.307 /docker-py/tests/integration/conftest.py:21: in setup_test_session
02:04:45.307     for data in c.pull(BUSYBOX, stream=True, decode=True):
02:04:45.307 /docker-py/docker/api/image.py:400: in pull
02:04:45.307     self._raise_for_status(response)
02:04:45.307 /docker-py/docker/api/client.py:258: in _raise_for_status
02:04:45.307     raise create_api_error_from_http_exception(e)
02:04:45.307 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:04:45.307     raise cls(e, response=response, explanation=explanation)
```

kolyshkin commented:
Fixed a nit from @thaJeztah; rebased to include #39243

cpuguy83 commented:
Haven't seen this one before

```
11:08:39 --- FAIL: TestCreateServiceSecretFileMode (8.89s)
11:08:39     create_test.go:224: Creating a new daemon
11:08:39     daemon.go:300: [d73753f11a294] waiting for daemon to start
11:08:39     daemon.go:332: [d73753f11a294] daemon started
11:08:39     create_test.go:265: assertion failed: 2 (int) != 1 (int)
11:08:39     daemon.go:284: [d73753f11a294] exiting daemon
```

kolyshkin commented:
> TestCreateServiceSecretFileMode

another flaky test :( https://github.com/moby/moby/issues/37132

kolyshkin commented:
experimental failure 
```
00:21:52.292 === RUN   TestServiceRemoveKeepsIngressNetwork
00:22:24.155 --- FAIL: TestServiceRemoveKeepsIngressNetwork (31.86s)
00:22:24.155     service_test.go:232: Creating a new daemon
00:22:24.155     daemon.go:300: [d36f5659ef00c] waiting for daemon to start
00:22:24.156     daemon.go:332: [d36f5659ef00c] daemon started
00:22:24.156     service_test.go:265: timeout hit after 30s: waiting for all tasks to be removed: task count at 1
00:22:24.156     daemon.go:284: [d36f5659ef00c] exiting daemon
```

thaJeztah commented:
@kolyshkin @cpuguy83 looks green now

kolyshkin commented:
Still need another review from @tonistiigi

kolyshkin commented:
...and @dmcgowan

dmcgowan reviewed (approved):
LGTM

thaJeztah reviewed (approved):
I'll follow @dmcgowan and @tonistiigi here; LGTM
